bson/datetime_ms.py

```diff
@@ -0,0 +1,134 @@
+import calendar
+import datetime
+import functools
+from typing import Any, Union, cast
+
+from bson.codec_options import (
+    DEFAULT_CODEC_OPTIONS,
+    CodecOptions,
+    DatetimeConversionOpts,
+)
+from bson.tz_util import utc
+
+EPOCH_AWARE = datetime.datetime.fromtimestamp(0, utc)
+EPOCH_NAIVE = datetime.datetime.utcfromtimestamp(0)
+
+
+class DatetimeMS:
+    """
+    Represents a BSON UTC datetime.
+
+    BSON UTC datetimes are defined as an int64 of milliseconds since the Unix
+    epoch. The principal use of DatetimeMS is to represent datetimes outside
+    the range of the Python builtin :class:`~datetime.datetime` class when
+    encoding/decoding BSON.
+
+    To decode UTC datetimes as a ``DatetimeMS``,`datetime_conversion` in
+    :class:`~bson.CodecOptions` must be set to 'datetime_ms' or
+    'datetime_auto'.
+    """
+
+    def __init__(self, value: Union[int, datetime.datetime]):
+        if isinstance(value, int):
+            self._value = value
+        elif isinstance(value, datetime.datetime):
+            self._value = _datetime_to_millis(value)
+        else:
+            raise TypeError(f"{type(value)} is not a valid type for DatetimeMS")
+
+    def __hash__(self) -> int:
+        return hash(self._value)
+
+    def __repr__(self) -> str:
+        return type(self).__name__ + "(" + str(self._value) + ")"
+
+    def __lt__(self, other: Union["DatetimeMS", int]) -> bool:
+        return self._value < other
+
+    def __le__(self, other: Union["DatetimeMS", int]) -> bool:
+        return self._value <= other
+
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, DatetimeMS):
+            return self._value == other._value
+        return False
+
+    def __ne__(self, other: Any) -> bool:
+        if isinstance(other, DatetimeMS):
+            return self._value != other._value
+        return True
+
+    def __gt__(self, other: Union["DatetimeMS", int]) -> bool:
+        return self._value > other
+
+    def __ge__(self, other: Union["DatetimeMS", int]) -> bool:
+        return self._value >= other
+
+    _type_marker = 9
+
+    def as_datetime(self, codec_options: CodecOptions = DEFAULT_CODEC_OPTIONS) -> datetime.datetime:
+        """Create a Python :class:`~datetime.datetime` from this DatetimeMS object.
+
+        :Parameters:
+          - `codec_options`: A CodecOptions instance for specifying how the
+            resulting DatetimeMS object will be formatted using ``tz_aware``
+            and ``tz_info``. Defaults to
+            :const:`~bson.codec_options.DEFAULT_CODEC_OPTIONS`.
+
+        .. versionadded:: 4.3
+        """
+        return cast(datetime.datetime, _millis_to_datetime(self._value, codec_options))
```

Comment:
Should we have a guard here to only allow DATETIME and DATETIME_CLAMP? Otherwise this could return a DatetimeMS, not a datetime. If you agree, let's open a new ticket in the epic for it rather than changing this PR.

Reply:
Yeah I think that makes sense, made an issue at [PYTHON-3365](https://jira.mongodb.org/browse/PYTHON-3365).